File: agent/pymada_client.js
const fs = require('fs');
const rp = require('request-promise');

if ('AGENT_PORT' in process.env){
    exports.host = "http://localhost:" + process.env['AGENT_PORT'];
} else {
    exports.host = "http://localhost:5001";
}

exports.getTask = async function(){
    const reqUrl = exports.host + '/get_task';
    let task_data = await rp({uri: reqUrl, method: 'POST', json: true});

    if (typeof(task_data['json_metadata']) == 'string'){
        try{
            task_data['json_metadata'] = JSON.parse(task_data['json_metadata']);
        } catch (e){
            if (!(e instanceof SyntaxError)){
                throw e;
            } 
        }
    }

    return task_data;
}

exports.saveResult = async function(result){
    const reqUrl = exports.host + '/save_results';
    const response = await rp({
        uri: reqUrl,
        method: 'POST',
        headers: {
            "Content-Type": "application/json"
        },
        body: result,
        json: true
    });
    return response;
}

exports.addUrl = async function(url, jsonMetadata=null){
    const reqUrl = exports.host + '/add_url';
    const response = await rp({
        uri: reqUrl,
        method: 'POST',
        headers: {
            "Content-Type": "application/json"
        },
        body: {'url': url, 'json_metadata': jsonMetadata},
        json: true
    });
    return response;
}

exports.logError = async function(errorMsg){
    const reqUrl = exports.host + '/log_error';
    const response = await rp({
        uri: reqUrl,
        method: 'POST',
        headers: {
            "Content-Type": "application/json"
        },
        body: {'message': errorMsg},
        json: true
    })
    return response;
}

exports.saveScreenshot = async function(screenshotPath){
    const reqUrl = exports.host + '/save_screenshot';
    const response = await rp({
        uri: reqUrl,
        method: 'POST',
        formData: {
            screenshot: fs.createReadStream(screenshotPath)
        },
        json: true
    });

    return response;
}

File: agent/pymada_client.py
import requests
import os

class Client(object):

    def __init__(self, host_url=None):
        if host_url is not None:
            self.host = host_url
            return

        if 'AGENT_PORT' in os.environ:
            self.host = 'http://localhost:' + os.environ['AGENT_PORT']

        self.host = 'http://localhost:5001'

    def get_task(self):
        req_url = self.host + '/get_task'
        r = requests.post(req_url)
        return r.json()
    
    def save_result(self, result):
        req_url = self.host + '/save_results'
        r = requests.post(req_url, json=result)
        return r.json()

    def add_url(self, url, json_metadata=None):
        req_url = self.host + '/add_url'
        r = requests.post(req_url, json={'url': url, 'json_metadata': json_metadata})
        return r.json()

    def log_error(self, err_msg):
        req_url = self.host + '/log_error'
        r = requests.post(req_url, json={'message': err_msg})
        return r.json()
    
    def save_screenshot(self, screenshot_path):
        req_url = self.host + '/save_screenshot'
        r = requests.post(req_url, files={'screenshot': open(screenshot_path, 'rb')})
        return r.json()